go-controller/pkg/ovn/master.go

```diff
@@ -1491,3 +1499,233 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
 	}
 	return nil
 }
+
+func (oc *Controller) addNodeObj(node *kapi.Node) error {
+	var gatewaysFailed sync.Map
+	var mgmtPortFailed sync.Map
+	var addNodeFailed sync.Map
+	var nodeClusterRouterPortFailed sync.Map
+	var errs []error
+
+	if noHostSubnet := noHostSubnet(node); noHostSubnet {
+		err := oc.lsManager.AddNoHostSubnetNode(node.Name)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	klog.Infof("Adding Node %q", node.Name)
+	hostSubnets, err := oc.addNode(node)
+	if err != nil {
+		klog.Errorf("NodeAdd: error creating subnet for node %s: %v", node.Name, err)
+		addNodeFailed.Store(node.Name, true)
+		nodeClusterRouterPortFailed.Store(node.Name, true)
+		mgmtPortFailed.Store(node.Name, true)
+		gatewaysFailed.Store(node.Name, true)
+		errs = append(errs, err)
+	}
+
+	if err = oc.syncNodeClusterRouterPort(node, hostSubnets); err != nil {
+		if !util.IsAnnotationNotSetError(err) {
+			errs = append(errs, err)
+		}
+		nodeClusterRouterPortFailed.Store(node.Name, true)
+	}
+
+	err = oc.syncNodeManagementPort(node, hostSubnets)
+	if err != nil {
+		if !util.IsAnnotationNotSetError(err) {
+			errs = append(errs, err)
+		}
+		mgmtPortFailed.Store(node.Name, true)
+	}
+
+	if err := oc.syncNodeGateway(node, hostSubnets); err != nil {
+		if !util.IsAnnotationNotSetError(err) {
+			errs = append(errs, err)
+		}
+		gatewaysFailed.Store(node.Name, true)
+	}
+
+	// ensure pods that already exist on this node have their logical ports created
+	options := metav1.ListOptions{FieldSelector: fields.OneTermEqualSelector("spec.nodeName", node.Name).String()}
+	pods, err := oc.client.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), options)
+	if err != nil {
+		errs = append(errs, err)
+	} else {
+		oc.addRetryPods(pods.Items)
+		oc.requestRetryPods()
+	}
+	if len(errs) == 0 {
+		return nil
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+func (oc *Controller) updateNodeObj(node, oldNode *kapi.Node) error {
+	var gatewaysFailed sync.Map
+	var mgmtPortFailed sync.Map
+	var addNodeFailed sync.Map
+	var nodeClusterRouterPortFailed sync.Map
+	var errs []error
+
+	shouldUpdate, err := shouldUpdate(node, oldNode)
+	if err != nil {
+		return err
+	}
+	if !shouldUpdate {
+		// the hostsubnet is not assigned by ovn-kubernetes
+		return err
+	}
+
+	var hostSubnets []*net.IPNet
+	_, failed := addNodeFailed.Load(node.Name)
+	if failed {
+		hostSubnets, err = oc.addNode(node)
+		if err != nil {
```

Comment:
should we add the node back to `addNodeFailed` in case it fails again to be added?

Reply:
the original code didn't added it and return so I kept the same behavior
```golang
if failed {
		hostSubnets, err = oc.addNode(node)
		if err != nil {
			klog.Errorf("NodeUpdate: error creating subnet for node %s: %v", node.Name, err)
			return
		}
		addNodeFailed.Delete(node.Name)
```